### PR TITLE
feat(toolsets/datadog): multi-account support for all four Datadog toolsets

### LINKED ISSUE #2065: Feature Request: Support multiple Datadog accounts in a single Holmes instance
### Problem

Many organizations run separate Datadog organizations for different environments (e.g. staging and production). Today, each Datadog toolset only supports a single set of credentials, so a Holmes instance can only query one Datadog org at a time. The only workarounds are:

- Run two separate Holmes deployments (one per org)
- Write brittle custom-toolset YAML to duplicate the built-in toolsets

### Proposed solution

Add an `accounts:` list to each Datadog toolset config, letting users declare N Datadog accounts. The LLM picks the target account per tool call via a new optional `account` parameter.

The existing single-account shorthand (`api_key` / `app_key` / `api_url` at the top level) keeps working unchanged — it is auto-normalised to a one-element account list named `default`.

**New multi-account config form:**

```yaml
toolsets:
  datadog/metrics:
    enabled: true
    config:
      accounts:
        - name: staging
          api_key: "{{ env.DD_API_KEY_STG }}"
          app_key: "{{ env.DD_APP_KEY_STG }}"
          api_url: https://api.datadoghq.eu
          default: true
        - name: production
          api_key: "{{ env.DD_API_KEY_PRD }}"
          app_key: "{{ env.DD_APP_KEY_PRD }}"
          api_url: https://api.datadoghq.eu
```

When two or more accounts are configured, the toolset's system prompt lists the available account names so the LLM can route calls correctly. The `account` parameter is exposed on every tool.

### Implementation

A PR is already open with a full implementation, tests, and docs: #2064

### PR DESCRIPTION
**Pull request by mdecalf**
Closes #2065

## Problem

Many organizations run separate Datadog organizations for staging and production. Until now a single Holmes instance could only query **one** Datadog org per toolset, which meant either running two Holmes deployments or writing brittle custom-toolset YAML. This PR adds first-class multi-account support to all four Datadog toolsets while keeping the existing single-account configuration entirely unchanged.

## Solution

A new `accounts:` list on each toolset config lets users declare N Datadog accounts. The LLM picks the target account on every tool call via a new optional `account` parameter. When only one account is configured (including all existing deployments), the parameter is a no-op — behavior is identical to today.

### Before (unchanged, still works)

```yaml
toolsets:
  datadog/metrics:
    enabled: true
    config:
      api_key: "{{ env.DD_API_KEY }}"
      app_key: "{{ env.DD_APP_KEY }}"
      api_url: https://api.datadoghq.com
```

### After (new multi-account form)

```yaml
toolsets:
  datadog/metrics:
    enabled: true
    config:
      accounts:
        - name: staging
          api_key: "{{ env.DD_API_KEY_STG }}"
          app_key: "{{ env.DD_APP_KEY_STG }}"
          api_url: https://api.datadoghq.eu
          default: true
        - name: production
          api_key: "{{ env.DD_API_KEY_PRD }}"
          app_key: "{{ env.DD_APP_KEY_PRD }}"
          api_url: https://api.datadoghq.eu
```

### `datadog_api.py`
- New `DatadogAccount` model holds per-account credentials (`name`, `api_key`, `app_key`, `api_url`, `default`). `api_url` defaults to `https://api.datadoghq.com` preserving the existing upstream default.
- `DatadogBaseConfig` gains `accounts: List[DatadogAccount]`. A `@model_validator` normalises the single-account shorthand into a one-element `[DatadogAccount(name="default", ...)]`, so zero migration is required.
- `_ui_required_fields = ["api_key", "app_key"]` keeps the UI form marking credentials as required; `_hidden_fields = ["accounts"]` hides the complex nested field from the form UI (advanced users configure it via YAML).
- `get_account(name)` resolves the per-call account name, or returns the default; raises `KeyError` listing available names so the LLM can self-correct on retry.
- `get_headers(account)` now takes a `DatadogAccount` instead of the full config.

### Four toolsets (logs, metrics, traces, general)
- Every tool gains an optional `account` string parameter.
- `_invoke` resolves the account via `get_account()` and passes it to `get_headers()` and all URL constructors.
- `_perform_healthcheck` takes a `DatadogAccount`; `prerequisites_callable` iterates all accounts and fails fast on the first unhealthy one, embedding the account name in the error message.
- `_reload_instructions()` passes the validated `dd_config` to the Jinja context so the system prompt lists available accounts and the default.

### Jinja instruction templates (all four)
Added a guard block listing configured accounts only when more than one is present. Single-account users see no change in their system prompt.

### `datadog_url_utils.py`
URL generators accept `DatadogAccount` instead of the full typed config, decoupling URL building from toolset config.

### Tests (67 pass, 16 skipped/slow)
- `tests/plugins/toolsets/datadog/test_datadog_accounts.py` (new, 11 tests): config parsing for both schemas, validation edge cases, and `get_account()` resolution.
- Account-routing tests added to all four toolset test files.
- Updated assertions in existing tests to match the new account-scoped healthcheck error messages.

### Docs
Added a "Multiple Datadog accounts" section to `docs/data-sources/builtin-toolsets/datadog.md`.

## Backward compatibility

| Scenario | Status |
|---|---|
| Existing `api_key`/`app_key`/`api_url` config | ✅ unchanged |
| Existing config without `api_url` (defaults to US1) | ✅ unchanged |
| Deprecated aliases (`dd_api_key`, `site_api_url`, …) | ✅ still work |
| Single toolset YAML, no `accounts:` key | ✅ auto-normalised |
| Multi-account `accounts:` list | ✅ new path |
| Both forms supplied together | ❌ `ValueError` (intentional) |

## Note on `llm_instructions` timing

`_reload_instructions()` is called from both `__init__` (renders with `dd_config=None`, template guard emits nothing) and from `prerequisites_callable` (renders with the validated config including the account list). Because `check_prerequisites` runs before the first LLM conversation in all deployment paths, the account list is always present in the system prompt when it matters.

## Summary by CodeRabbit

* **New Features**
  * Datadog toolset: configure multiple accounts (with legacy single-account shorthand preserved), per-request account parameter, per-account health checks; omitting account uses the configured/default account.

* **Documentation**
  * Updated usage and troubleshooting instructions to list configured accounts, mark the default, and show how to target a specific account.

* **Tests**
  * Added tests for multi-account configs, resolver behavior, per-account routing, and healthcheck/error messaging across logs, metrics, traces, and general toolsets.

[Review Change Stack](https://app.coderabbit.ai/change-stack/HolmesGPT/holmesgpt/pull/2064?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: holmes/plugins/toolsets/datadog/datadog_api.py [hunk 1]
```diff
@@ -95,33 +95,120 @@ def convert_api_url_to_app_url(api_url: Union[str, AnyUrl]) -> str:
...
+    api_key: str = Field(
+        title="API Key",
+        description="Datadog API key for authentication",
+        json_schema_extra={"format": "password"},
+    )
+    app_key: str = Field(
+        title="Application Key",
+        description="Datadog application key for authentication",
+        json_schema_extra={"format": "password"},
+    )
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/datadog_api.py:129]: fixed

File: holmes/plugins/toolsets/datadog/datadog_url_utils.py [hunk 2]
```diff
@@ -118,8 +116,8 @@ def generate_datadog_logs_url(
         "storage": params["filter"]["storage_tier"],
     }
 
-    if dd_config.indexes != ["*"]:
-        url_params["index"] = ",".join(dd_config.indexes)
+    if indexes != ["*"]:
+        url_params["index"] = ",".join(indexes)
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/datadog_url_utils.py]: fixed

File: holmes/plugins/toolsets/datadog/datadog_api.py [hunk 3]
```diff
@@ -132,12 +221,104 @@ class DatadogBaseConfig(ToolsetConfig):
...
+                )
+            self.accounts = [
+                DatadogAccount(
+                    name="default",
+                    api_key=self.api_key,  # type: ignore[arg-type]
+                    app_key=self.app_key,  # type: ignore[arg-type]
+                    api_url=self.api_url,  # type: ignore[arg-type]
+                    default=True,
+                )
+            ]
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/datadog_api.py]: fixed

File: holmes/plugins/toolsets/datadog/toolset_datadog_metrics.py [hunk 4]
```diff
@@ -766,18 +841,40 @@ def prerequisites_callable(self, config: dict[str, Any]) -> Tuple[bool, str]:
...
     def _reload_instructions(self):
-        """Load Datadog metrics specific troubleshooting instructions."""
+        """Load Datadog metrics specific troubleshooting instructions.
+
+        Passes the validated ``dd_config`` (with its accounts list) to the
+        Jinja context so the template can surface the available account
+        names to the LLM. ``dd_config`` is None on first render (called
+        from ``__init__``) and the template guards against that.
+        """
+        from holmes.plugins.prompts import load_and_render_prompt
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/toolset_datadog_metrics.py]: fixed

**Comment by mdecalf** [comment]
@coderabbitai review

**Comment by aantn** [comment]
@claude review

File: holmes/plugins/toolsets/datadog/datadog_api.py [hunk 5]
```diff
@@ -95,33 +95,124 @@ def convert_api_url_to_app_url(api_url: Union[str, AnyUrl]) -> str:
...
-
-    api_key: str = Field(
+    # The UI form guides users through the common single-account path, so mark
+    # the shorthand credential fields as required in the exported schema even
+    # though they are Optional at the Pydantic level (to allow the
+    # `accounts:` alternative).  The `accounts` list is hidden because nested
+    # complex objects don't render well in form UIs — advanced users supply it
+    # directly in their YAML/Helm values.
+    _ui_required_fields: ClassVar[List[str]] = ["api_key", "app_key"]
+    _hidden_fields: ClassVar[List[str]] = ["accounts"]
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/datadog_api.py:184]: Fixed in 55be47420. Went with the subclass-extension approach: both `DatadogTracesConfig` and `DatadogLogsConfig` now declare `_hidden_fields = ["accounts", "indexes"]` explicitly, so the parent's `"accounts"` is never dropped by ClassVar shadowing. Skipped the MRO-walking alternative in `pydantic_utils.py` to keep the change minimal and localised.

File: holmes/plugins/toolsets/datadog/toolset_datadog_logs.py [hunk 6]
```diff
@@ -133,19 +149,23 @@ def prerequisites_callable(self, config: dict[str, Any]) -> Tuple[bool, str]:
             dd_config = DatadogLogsConfig(**config)
             self.dd_config = dd_config
 
-            success, error_msg = self._perform_healthcheck()
-            return success, error_msg
+            for account in dd_config.accounts:
+                success, error_msg = self._perform_healthcheck(account)
+                if not success:
+                    return False, error_msg
+            self._reload_instructions()
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/toolset_datadog_logs.py:157]: Fixed in 55be47420. `_perform_healthcheck` now takes `dd_config: DatadogLogsConfig` as an explicit parameter (reads `dd_config.indexes` / `dd_config.timeout_seconds` directly instead of `self.dd_config`), the `if not self.dd_config` guard removed, and `self.dd_config = dd_config` moved to after the per-account loop — identical pattern to metrics and traces.

File: holmes/plugins/toolsets/datadog/toolset_datadog_metrics.py [hunk 7]
```diff
@@ -110,8 +140,8 @@ def _invoke(self, params: dict, context: ToolInvokeContext) -> StructuredToolRes
                 default_time_span_seconds=ACTIVE_METRICS_DEFAULT_TIME_SPAN_SECONDS,
             )
 
-            url = f"{self.toolset.dd_config.api_url}/api/v1/metrics"
-            headers = get_headers(self.toolset.dd_config)
+            url = f"{account.api_url}/api/v1/metrics"
```
[the remark this answers is not in the document]
Comment (mdecalf) [reply holmes/plugins/toolsets/datadog/toolset_datadog_metrics.py]: Fixed in 55be47420. Applied `str(account.api_url).rstrip("/")` at all 10 URL construction sites across logs (2), metrics (5) and traces (3), consistent with the pattern already used in `toolset_datadog_general.py`. Went with the call-site approach rather than a `DatadogAccount` validator to avoid Pydantic v2 `AnyUrl` type reconstruction complexity.

**Comment by mdecalf** [comment]
@aantn Following PR #2115 I guess we can close this one
